tfhe/src/shortint/server_key/tests/noise_level.rs

```diff
@@ -0,0 +1,263 @@
+use crate::shortint::ciphertext::NoiseLevel;
+use crate::shortint::keycache::KEY_CACHE;
+use crate::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS;
+use crate::shortint::{Ciphertext, ServerKey};
+
+fn test_1_ct_noise_level_propagation(sk: &ServerKey, ct: &Ciphertext) {
+    let test_fn = |f: &dyn Fn(&ServerKey, &Ciphertext) -> Ciphertext,
+                   g: &dyn Fn(NoiseLevel) -> NoiseLevel| {
+        assert!(f(sk, ct).noise_level == g(ct.noise_level));
+    };
+
+    test_fn(&ServerKey::unchecked_neg, &|a| a);
+    test_fn(
+        &|sk, ct| ServerKey::unchecked_neg_with_correcting_term(sk, ct).0,
+        &|a| a,
+    );
+
+    let acc = sk.generate_lookup_table(|_| 0);
+
+    test_fn(
+        &|sk, ct| ServerKey::apply_lookup_table(sk, ct, &acc),
+        &|_| NoiseLevel::NOMINAL,
+    );
+}
+
+fn test_1_ct_assign_noise_level_propagation(sk: &ServerKey, ct: &Ciphertext) {
+    let test_fn = |f: &dyn Fn(&ServerKey, &mut Ciphertext),
+                   g: &dyn Fn(NoiseLevel) -> NoiseLevel| {
+        let mut clone = ct.clone();
+        f(sk, &mut clone);
+        assert!(clone.noise_level == g(ct.noise_level));
+    };
+
+    test_fn(&ServerKey::unchecked_neg_assign, &|a| a);
+    test_fn(
+        &|sk, ct| {
+            ServerKey::unchecked_neg_assign_with_correcting_term(sk, ct);
+        },
+        &|a| a,
+    );
+
+    let acc = sk.generate_lookup_table(|_| 0);
+
+    test_fn(
+        &|sk, ct| ServerKey::apply_lookup_table_assign(sk, ct, &acc),
+        &|_| NoiseLevel::NOMINAL,
+    );
+}
+
+fn test_2_ct_noise_level_propagation(sk: &ServerKey, ct1: &Ciphertext, ct2: &Ciphertext) {
+    let test_fn = |f: &dyn Fn(&ServerKey, &Ciphertext, &Ciphertext) -> Ciphertext,
+                   g: &dyn Fn(NoiseLevel, NoiseLevel) -> NoiseLevel| {
+        assert!(f(sk, ct1, ct2).noise_level == g(ct1.noise_level, ct2.noise_level));
+    };
```

Comment:
can this be made a function with #[track_caller] so that we have the test triggering the error if a test fails ? unless closures can be track_caller bu I don't think it's the case, to combine with the naming improvement for f and g proposed earlier in the file and applied to all similar test_fn

Reply:
Does not seem to work on closures

Reply:
can't the test_fn be defined like

```rust
#[track_caller]
fn test_fn(...) {
}
```
 ?

Reply:
To keep the rest of the code simple (not passing sk, ct1, ct2 explicitely everywhere), `test_fn` (as you just defined it) would have to be wrapped in a closure

Reply:
Ok should be fine with backtrace anyways